SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp

```diff
@@ -44,27 +44,68 @@ EntityFactory::EntityFactory(USpatialNetDriver* InNetDriver, USpatialPackageMapC
 {
 }
 
-TArray<FWorkerComponentData> EntityFactory::CreateEntityComponents(USpatialActorChannel* Channel,
-																   FRPCsOnEntityCreationMap& OutgoingOnCreateEntityRPCs,
-																   uint32& OutBytesWritten)
+EntityComponents EntityFactory::CreateSkeletonEntityComponents(AActor* Actor)
 {
-	AActor* Actor = Channel->Actor;
 	UClass* Class = Actor->GetClass();
-	Worker_EntityId EntityId = Channel->GetEntityId();
 
-	FString ClientWorkerAttribute = GetConnectionOwningWorkerId(Actor);
+	WorkerRequirementSet ReadAcl;
+	if (Class->HasAnySpatialClassFlags(SPATIALCLASS_ServerOnly))
+	{
+		ReadAcl = SpatialConstants::UnrealServerPermission;
+	}
+	else
+	{
+		ReadAcl = SpatialConstants::ClientOrServerPermission;
+	}
 
-	WorkerRequirementSet AnyServerRequirementSet = { SpatialConstants::UnrealServerAttributeSet };
-	WorkerRequirementSet AnyServerOrClientRequirementSet = { SpatialConstants::UnrealServerAttributeSet,
-															 SpatialConstants::UnrealClientAttributeSet };
+	WriteAclMap ComponentWriteAcl;
+	ComponentWriteAcl.Add(SpatialConstants::ENTITY_ACL_COMPONENT_ID, SpatialConstants::UnrealServerPermission);
+
+	EntityComponents EntityComps;
+	EntityComps.ModifiableComponents.Add(Position::ComponentId,
+										 MakeUnique<Position>(Coordinates::FromFVector(GetActorSpatialPosition(Actor))));
+	EntityComps.ModifiableComponents.Add(Metadata::ComponentId, MakeUnique<Metadata>(Class->GetName()));
+	EntityComps.ModifiableComponents.Add(SpawnData::ComponentId, MakeUnique<SpawnData>(Actor));
+
+	if (ShouldActorHaveVisibleComponent(Actor))
+	{
+		EntityComps.ComponentDatas.Add(ComponentFactory::CreateEmptyComponentData(SpatialConstants::VISIBLE_COMPONENT_ID));
+	}
+
+	if (!Class->HasAnySpatialClassFlags(SPATIALCLASS_NotPersistent))
+	{
+		EntityComps.ComponentDatas.Add(Persistence().CreateComponentData());
+	}
+
+	if (Actor->NetDormancy >= DORM_DormantAll)
+	{
+		EntityComps.ComponentDatas.Add(ComponentFactory::CreateEmptyComponentData(SpatialConstants::DORMANT_COMPONENT_ID));
+	}
+
+	if (Actor->IsA<APlayerController>())
+	{
+		EntityComps.ComponentDatas.Add(Heartbeat().CreateComponentData());
+	}
+
+	EntityComps.ModifiableComponents.Add(EntityAcl::ComponentId, MakeUnique<EntityAcl>(ReadAcl, ComponentWriteAcl));
+
+	// Add Actor completeness tags.
+	EntityComps.ComponentDatas.Add(ComponentFactory::CreateEmptyComponentData(SpatialConstants::LB_TAG_COMPONENT_ID));
+
+	return EntityComps;
+}
+
+void EntityFactory::WriteLBComponents(EntityComponents& EntityComps, AActor* Actor)
```

Comment:
I think if I tried more, I could make this function static, and thus make its dependencies explicit.
Worth?

Reply:
no, this is the factory, makes some sense to keep the state the factory needs in the factory and only take required arguments in the function (but if this could be static, could you make the function const? Or does it have side effects on the factory)